### ISSUE TITLE
 The issue about unexpected inference result (output class and probability-value) through quantized tflite model in more than a certain number of classes

### ISSUE THREAD

**Issue by pdkz** [tensorflow/tensorflow#43700]
**System information**
- Model training (VGG16) environment
    - OS Platform and Distribution
        - Windows 10
    - GPU
        - NVIDIA GTX1080
    - Cuda Toolkit 10.0.130
    - Python 3.7
    - Keras 2.2.4
    - TensorFlow 1.14.0

- Model training (MobileNetV2) environment    
    - OS Platform and Distribution
        - Ubuntu 18.04.4 LTS (GNU/Linux 4.15.0-99-generic x86_64)
    - GPU
        - NVIDIA Quadro P5000
    - Cuda Toolkit 10.1.243
    - Python 3.7
    - Keras 2.2.4
    - TensorFlow 1.14.0

- Model convert environment
    - OS Platform and Distribution
        - Mac OS X 10.14.5 (18F132)
    - Python 3.6.7
    - Keras 2.2.4
    - TensorFlow 1.14.0

- Target mobile device
    - Pixel 3 XL
        - Android 9 (build PQ3A.1980801.002)

**Trained model information**
- VGG16 and MobileNetV2 model (transfer learning)
- The number of classes: **10,996**
    - This model is for classification of  the huge classes of the images of a certain category.

**Abstract of the issue**
- In case did the tflite conversion with enable quantization option for this model, the output class (1st rank) for the input test data(about 1000-2000 jpg images files) to 10,996 class model was completely different all from the class of input test data, and I got the unexpected output (the output class did not match in all test data and the probability-value in 1st rank is about 0.4-0.5) compared with the output for 10,806 class model.
- But in disabled quantization option, the output class matched the class of the input test data, the probability-value in 1st rank is 0.9 and more, and the classification accuracy for the same test data was 90% and more.
- **This issue did not occur in the model of less 10,806 classes**
    - I'm investigating if this issue occurs between 10,806 and 10,996 classes.
- This issue occurred on mobile device (Pixel 3 XL) but similarly reproduced on Mac too.
- The method of the quantization is dynamic range quantization.

**Output example**
- Classes
    - [ a_0, a_1, ... , a_10995 ]
- Input test data class (expected output class)
    - a_1234
- Output class and probability-values **(Quantized)**
    ```
    a_4952 0.49650624
    a_5831 0.044097286
    a_6422 0.032513995
    ...
    ```
-  Output class and probability-values **(Not-quantized)**
    ```
    a_1234 0.92802435
    a_1235 0.004674452
    a_13731 0.0012423205
    ...
    ```

**Question**
- Is there something that is considered to be the cause of this issue?
- In case a model has the huge number of classes, in internal calculation of a model, overflow or zero-division occurred by applying dynamic range quantization to a trained model, and as a result, is there a possibility that the output result is something wrong?

- A similar issue occurred in CoreML:
    - In CoreML, when a model has the model more than 9,000 classes, this resulted in overflow with Neural Engine mode and GPU mode, but I could avoid this issue with enable CPU mode. 
- In Android, I couldn't avoid this issue with the same method.

**Source code for model conversion**
```python
import os
import tensorflow as tf
from keras.models import load_model, model_from_json
from keras.optimizers import SGD

def convert(input_path, 
            weights_file,
            output_dir, 
            output_filename, 
            initial_lr=2e-4, 
            momentum=0.9, 
            is_quantized=False):

    model = model_from_json(open(os.path.join(input_path, 'model', 'model.json')).read())
    model.load_weights(os.path.join(input_path, 'model', weights_file))
    model.summary()

    opt = SGD(initial_lr, momentum=momentum)
    model.compile(
        optimizer=opt,
        loss='categorical_crossentropy'
    )

    output_modeldir = os.path.split(input_path)[-1]
    output_dir = os.path.join(output_dir, 'models', output_modeldir)

    os.makedirs(output_dir, exist_ok=True)

    keras_file = os.path.join(output_dir, 'weights_temp.h5')
    model_ = model.save(keras_file)

    converter = tf.lite.TFLiteConverter.from_keras_model_file(keras_file)

    if is_quantized:
        converter.post_training_quantize=True
        output_filename = f'{output_filename}_quantization.tflite'
    else:
        output_filename = f'{output_filename}.tflite'

    tflite_model = converter.convert()

    tflite_file = os.path.join(output_dir, f'{output_filename}')
    with open(tflite_file, 'wb') as f:
        f.write(tflite_model)

    print(f'Saved [{tflite_file}]')
```

**Output logs**

```
Using TensorFlow backend.
WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3976: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:4185: The name tf.truncated_normal is deprecated. Please use tf.random.truncated_normal instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:131: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:133: The name tf.placeholder_with_default is deprecated. Please use tf.compat.v1.placeholder_with_default instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
2020-10-01 11:34:43.751852: I tensorflow/core/platform/cpu_feature_guard.cc:145] This TensorFlow binary is optimized with Intel(R) MKL-DNN to use the following CPU instructions in performance critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in non-MKL-DNN operations, rebuild TensorFlow with the appropriate compiler flags.
2020-10-01 11:34:43.752602: I tensorflow/core/common_runtime/process_util.cc:115] Creating new thread pool with default inter op setting: 6. Tune using inter_op_parallelism_threads for best performance.
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
image_input (InputLayer)     (None, 128, 128, 3)       0
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 128, 128, 64)      1792
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 128, 128, 64)      36928
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 64, 64, 64)        0
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 64, 64, 128)       73856
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 64, 64, 128)       147584
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, 32, 32, 128)       0
_________________________________________________________________
block3_conv1 (Conv2D)        (None, 32, 32, 256)       295168
_________________________________________________________________
block3_conv2 (Conv2D)        (None, 32, 32, 256)       590080
_________________________________________________________________
block3_conv3 (Conv2D)        (None, 32, 32, 256)       590080
_________________________________________________________________
block3_pool (MaxPooling2D)   (None, 16, 16, 256)       0
_________________________________________________________________
block4_conv1 (Conv2D)        (None, 16, 16, 512)       1180160
_________________________________________________________________
block4_conv2 (Conv2D)        (None, 16, 16, 512)       2359808
_________________________________________________________________
block4_conv3 (Conv2D)        (None, 16, 16, 512)       2359808
_________________________________________________________________
block4_pool (MaxPooling2D)   (None, 8, 8, 512)         0
_________________________________________________________________
block5_conv1 (Conv2D)        (None, 8, 8, 512)         2359808
_________________________________________________________________
block5_conv2 (Conv2D)        (None, 8, 8, 512)         2359808
_________________________________________________________________
block5_conv3 (Conv2D)        (None, 8, 8, 512)         2359808
_________________________________________________________________
block5_pool (MaxPooling2D)   (None, 4, 4, 512)         0
_________________________________________________________________
gap_vgg16gap (GlobalAverageP (None, 512)               0
_________________________________________________________________
fc1_vgg16gap (Dense)         (None, 4096)              2097152
_________________________________________________________________
bn1_vgg16gap (BatchNormaliza (None, 4096)              16384
_________________________________________________________________
relu1_vgg16gap (Activation)  (None, 4096)              0
_________________________________________________________________
do1_vgg16gap (Dropout)       (None, 4096)              0
_________________________________________________________________
fc2_vgg16gap (Dense)         (None, 4096)              16777216
_________________________________________________________________
bn2_vgg16gap (BatchNormaliza (None, 4096)              16384
_________________________________________________________________
relu2_vgg16gap (Activation)  (None, 4096)              0
_________________________________________________________________
do2_vgg16gap (Dropout)       (None, 4096)              0
_________________________________________________________________
predictions_vgg16gap (Dense) (None, 10996)             45050612
=================================================================
Total params: 78,672,436
Trainable params: 71,020,788
Non-trainable params: 7,651,648
_________________________________________________________________
WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py:507: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with distribution=normal is deprecated and will be removed in a future version.
Instructions for updating:
`normal` is a deprecated alias for `truncated_normal`
2020-10-01 11:34:49.269923: I tensorflow/core/grappler/devices.cc:60] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0 (Note: TensorFlow was not compiled with CUDA support)
2020-10-01 11:34:49.270019: I tensorflow/core/grappler/clusters/single_machine.cc:359] Starting new session
2020-10-01 11:34:49.759721: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:716] Optimization results for grappler item: graph_to_optimize
2020-10-01 11:34:49.759743: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:718]   function_optimizer: function_optimizer did nothing. time = 0.007ms.
2020-10-01 11:34:49.759750: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:718]   function_optimizer: function_optimizer did nothing. time = 0.004ms.
WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/tensorflow/lite/python/util.py:238: convert_variables_to_constants (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.compat.v1.graph_util.convert_variables_to_constants`
WARNING:tensorflow:From /anaconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/graph_util_impl.py:270: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.compat.v1.graph_util.extract_sub_graph`
/anaconda3/envs/py36/lib/python3.6/site-packages/tensorflow/lite/python/lite.py:769: UserWarning: Property post_training_quantize is deprecated, please use optimizations=[Optimize.DEFAULT] instead.
  " instead." % name)
2020-10-01 11:34:50.957910: I tensorflow/core/grappler/devices.cc:60] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0 (Note: TensorFlow was not compiled with CUDA support)
2020-10-01 11:34:50.958019: I tensorflow/core/grappler/clusters/single_machine.cc:359] Starting new session
2020-10-01 11:34:51.559693: E tensorflow/core/grappler/grappler_item_builder.cc:637] Init node block1_conv1/kernel/Assign doesn't exist in graph
Saved [./weights_quantization.tflite]
```

**Comment by pdkz** [comment]
Hi all, Is there an anything update a progressing about the solution for this issue?
Thanks.

**Comment by jaeyoo** [comment]
Hi @pdkz I am still working on it. Please stay tuned. thanks!

**Comment by jaeyoo** [comment]
@pdkz Hi pdkz, Sorry for belated reply. Could you please check the conversion with tf-nightly >= 2.4.0-dev20201001? I am still investigating, but I want to check it in parallel on both sides. Thanks.

**Comment by pdkz** [comment]
Hi @jaeyoo Never mind, I'm grateful for your support. 
Okay, I'll try it in the model conversion using tf-nightly 2.4.0 greater. I'd like to be helpful to you even just a little.
Thanks.

**Comment by pdkz** [comment]
Hi @jaeyoo!
I tried to do the conversion in tf-nightly 2.5.0.
At the moment, it seems that the converted model output the expected output when I checked the accuracy. Thanks for your advice and supports!
But it's not clear what's caused it, so I'm also investigating the source code to find out the cause for this issue.
Is this issue known or latent bug in TF 1.x ? (It seems I couldn't find posts about it in issue tracker. )
Please tell me if you find out the cause or the hint.

**Comment by jaeyoo** [comment]
Hi @pdkz Thank you for sharing good news! Yes... honestly speaking, I couldn't reproduce the result at the latest TF as you saw, and I tried to dig into where the cause was, but I couldn't find it yet. also, it wasn't known issue, so I was surprised at the first glance at this issue. Please let me inform you if I can find it. Thank you!
